### PR TITLE
Add ReduceContext, ReduceError, Example tests

### PR DESCRIPTION
**Pull request by chrispappas**

### CODE REVIEW

File: utils/reduce.go [hunk 1]
```diff
@@ -13,6 +14,36 @@ func Reduce[T, Acc any](
...
 
+// ReduceContext passes the context arg through to the reducer fn.
+func ReduceContext[T, Acc any](
+	ctx context.Context,
+	arr []T,
+	fn func(ctx context.Context, acc Acc, elem T, index int) Acc,
+	acc Acc,
+) Acc {
+	for i, elem := range arr {
+		acc = fn(ctx, acc, elem, i)
```
[the remark this answers is not in the document]
Comment (chris-langager) [reply utils/reduce.go:25]: Not exactly picking a side here, but I can say that for this one, we are using the index.  The use case we have is when processing an array of input that just came over the wire and performing validation on it.  We reduce over all of the different input values to a single error (or nil) containing information about what was wrong (ie. missing required field or whatever). While building up the error message, we use the index in the message to let the user know "Hey, the object at index 5 had a validation error)

[the remark this answers is not in the document]
Comment (chris-langager) [reply utils/reduce.go:25]: If index was not in the function signature, there's a few other ways we could do this while still using reuse.  Just wanted to share that to see if it sparked any opinions one way or the other.

[the remark this answers is not in the document]
Comment (chrispappas) [reply utils/reduce.go:25]: I'm ok with leaving it and ignoring it when it's not needed... easy to just add `, _ int` to the fn signature

[the remark this answers is not in the document]
Comment (cuducos) [reply utils/reduce.go:25]: Awesome. So I think I'll revert that in #6 (maybe in a follow up PR, so we get that one going).

[the remark this answers is not in the document]
Comment (cuducos) [reply utils/reduce.go:25]: From the snippet above, my understanding is that:

* we're using the index only to produce error messages
* we can wrap the error message in the `MapConcurent` to include the index in the error message instead

But… we can take this in small steps. I'll revert for consistency and flexibility!

**Pull request merged**